test: Added unit tests for backend controllers using jest

hemant14050 opened the pull request:
## Summary
This pull request (PR) aims to enhance the project by adding unit tests for the backend controllers and improving the overall test coverage.

## Description
In this PR, 
 - Improved project structure by moving controllers from the `routes` folder to a dedicated `controllers` folder. The existing controller functions have been moved from the routes folder to this new location. This change enhances the overall code organization and maintainability of the application.
 - created setup for tests using jest
 - Added unit tests for various backend controller functions

## Images
- Code Coverage
![image](https://github.com/krishnaacharyaa/wanderlust/assets/112842857/b8b888a7-fb2c-4f3f-b9a4-a3bea8feb3c6)

## Issue(s) Addressed
 **Closes #5**

## Prerequisites

- [ ] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions) ?

krishnaacharyaa commented:
Hi @hemant14050, Thank you for your efforts and time for writing the much needed test cases, greatly appreciated. One small request, can you kindly rebase your branch and commit, because we had recent changes merged in the posts api...

krishnaacharyaa commented on backend/controllers/postsController.js:
Kindly rename the file to post-controller.js

krishnaacharyaa commented on backend/tests/unit/controllers/postsController.test.js:
File name here aswell :)

krishnaacharyaa reviewed (changes requested):
Thank you for your contribution, its very useful for the project, Kindly follow through couple of changes

hemant14050 commented:
> Hi @hemant14050, Thank you for your efforts and time for writing the much needed test cases, greatly appreciated. One small request, can you kindly rebase your branch and commit, because we had recent changes merged in the posts api...

Hello @krishnaacharyaa, I have rebased and added new changes to posts api. Please review and update me if any other changes need from my side :)

chinmaykunkikar commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,312 @@
+    const postObject = {
+      title: 'Test Post',
+      authorName: 'Test Author',
+      imageLink: 'https://www.forTestingPurposeOnly.com',
```
A thing that caught my eye here:

The test should actually fail with this `imageLink`. This URL won't be accepted on the frontend because there's a regex that checks if the image URL ends with .png/.jpg/.jpeg.
See this line for reference: https://github.com/krishnaacharyaa/wanderlust/blob/2208b82b94f0932ec5cb6bfde1f62487f7881b8b/frontend/src/pages/add-blog.tsx#L63

Now, I get that the test will pass nevertheless because there is no check for this in the API handlers or in the model logic. We should open an issue for this and resolve it.

chinmaykunkikar replied:
Feel free to mark it resolved after opening an issue @hemant14050 or @krishnaacharyaa.

krishnaacharyaa replied:
That's a very nice catch @chinmaykunkikar, according to me we can solve it here in this PR as this is still not deployed in production, had it been already deployed we could have given fix for this by raising ticket, as it is other-way around we can fix it in this very PR, @hemant14050 kindly note the comments and make required changes,

For us the target is 3 checks of https://github.com/krishnaacharyaa/wanderlust/pull/48#pullrequestreview-1716115968, post these checks we can merge, otherwise it defeats the very purpose of adding test cases :)

chinmaykunkikar reviewed (commented):
The tests look good to me.
It's just that we need to do some checks at the backend to make sure that -
1. URLs without image extensions shouldn't be accepted in the database. The API handlers should take care of this. (https://github.com/krishnaacharyaa/wanderlust/blob/2208b82b94f0932ec5cb6bfde1f62487f7881b8b/frontend/src/pages/add-blog.tsx#L63)
2. The API should also fail if all form fields are not filled (https://github.com/krishnaacharyaa/wanderlust/blob/2208b82b94f0932ec5cb6bfde1f62487f7881b8b/frontend/src/pages/add-blog.tsx#L57)
3. Backend should handle a situation if more than 3 categories are picked (https://github.com/krishnaacharyaa/wanderlust/blob/2208b82b94f0932ec5cb6bfde1f62487f7881b8b/frontend/src/pages/add-blog.tsx#L67).

These checks are done on the frontend but they should match the API response on the backend as well. It's a good practice in case someone tried to bypass these limits of the frontend by directly modifying API requests using Postman or some other API testing tools, or while API testing like this.

chinmaykunkikar commented:
We can -
1. convert this PR into a draft PR and work on these issues.
OR
2. or merge this PR for now and add new tests when we're done working on those issues.

chinmaykunkikar commented:
Hey @hemant14050, can I indulge you into something that bothers me about this PR? 😛 
This can actually be split into two separate PRs, as there are two features being merged in this PR:
1. The test setup and the actual tests.
2. The API controllers (just the first commit of this PR: 349e0f29c667ef9a31731ae57a07253641064ea5)

The controllers are good to go, but, because the tests potentially need some work, the controllers are also being blocked along with tests.

Let's forget about it this time... or if you're up for it, I can help you with creating a separate PR for the controllers. Totally your choice :)

krishnaacharyaa commented:
> Hey @hemant14050, can I indulge you into something that bothers me about this PR? 😛 This can actually be split into two separate PRs, as there are two features being merged in this PR:
> 
> 1. The test setup and the actual tests.
> 2. The API controllers (just the first commit of this PR: [349e0f2](https://github.com/krishnaacharyaa/wanderlust/commit/349e0f29c667ef9a31731ae57a07253641064ea5))
> 
> The controllers are good to go, but, because the tests potentially need some work, the controllers are also being blocked along with tests.
> 
> Let's forget about it this time... or if you're up for it, I can help you with creating a separate PR for the controllers. And that's totally your choice :)

That's seems very reasonable to me, @hemant14050 can we do it this way, so that your efforts are considered and pushed as a PR and you can then work on this test cases :) For setting up the PR from this kindly take help from @chinmaykunkikar, he would be happy to help you if you face any difficulties ...
